cmd/chunks-inspect/loki.go

```diff
@@ -122,27 +129,96 @@ func parseLokiChunk(chunkHeader *ChunkHeader, r io.Reader) (*LokiChunk, error) {
 		return nil, fmt.Errorf("failed to read rawData for Loki chunk into memory: %w", err)
 	}
 
+	//Magic Number
 	if num := binary.BigEndian.Uint32(data[0:4]); num != 0x012EE56A {
 		return nil, fmt.Errorf("invalid magic number: %0x", num)
 	}
 
 	// Chunk format is at position 4
 	f := data[4]
 
+	// Compression
 	compression, err := getCompression(f, data[5])
 	if err != nil {
 		return nil, fmt.Errorf("failed to read compression: %w", err)
 	}
 
 	// return &LokiChunk{encoding: compression}, nil
 
-	metasOffset := binary.BigEndian.Uint64(data[len(data)-8:])
+	// This was copied from memchunk.go newByteChunk() as a helper function to use here also
+	// In format >=v4 there are multiple metadata sections at the end of the chunk file,
+	// the offset and length of each is stored as the last bytes in the file
+	readSectionLenAndOffset := func(idx int) (uint64, uint64) {
+		lenAndOffsetPos := len(data) - (idx * 16)
+		lenAndOffset := data[lenAndOffsetPos : lenAndOffsetPos+16]
+		return binary.BigEndian.Uint64(lenAndOffset[:8]), binary.BigEndian.Uint64(lenAndOffset[8:])
+	}
 
-	metadata := data[metasOffset : len(data)-(8+4)]
+	// Chunk formats v1-v3 had a single metadata section which was stored at the end of the chunk with the last 8 bytes a pointer to the beginning of the metadata table section
+	metasOffset := uint64(0)
+	metasLen := uint64(0)
+	if f < chunkFormatV4 {
+		// Metadata table is at the end, read the last 8 bytes to get the offset to the start of the table
+		metasOffset = binary.BigEndian.Uint64(data[len(data)-8:])
+		// Exclude the last 8 bytes which is the offset, and the 4 bytes before that which is the checksum from the length
+		metasLen = uint64(len(data)-(8+4)) - metasOffset
+	} else {
+		// the chunk metas section is index 1
+		metasLen, metasOffset = readSectionLenAndOffset(1)
+	}
 
-	metaChecksum := binary.BigEndian.Uint32(data[len(data)-12 : len(data)-8])
+	// Read metadata
+	metadata := data[metasOffset : metasOffset+metasLen]
+
+	// Read metadata block checksum, the last 4 bytes before the last 8 bytes containing the metadata table start offset
+	metaChecksum := binary.BigEndian.Uint32(data[metasOffset+metasLen:])
 	computedMetaChecksum := crc32.Checksum(metadata, castagnoliTable)
 
+	// If chunkFormat >= v4 we also need to read the structured metadata section
+	var structuredMetadataSymbols []string
+	if f >= chunkFormatV4 {
+		// the chunk metas section is index 2
+		structuredMetadataLength, structuredMetadataOffset := readSectionLenAndOffset(2)
+
+		//expCRC := binary.BigEndian.Uint32(data[structuredMetadataOffset+structuredMetadataLength:])
+		//computedMetaChecksum := crc32.Checksum(lb, castagnoliTable)
+
+		structuredMetadata := data[structuredMetadataOffset : structuredMetadataOffset+structuredMetadataLength]
+
+		// Structured Metadata is "normalized" or "compressed" by storing an index to a string with each log line, and then all the strings in the chunk metadata section
```

Comment:
nit: By reading the code not sure I understand the "normalized" part of this comment, but maybe I'm just missing context